### PR TITLE
Reflect spec changes in delegate immediate operand

### LINKED ISSUE #176: Inconsistency in documentation for try-delegate behavior for function body
In #175 there's some discussion of what the intended behavior of `try-delegate` with a label pointing to the function body is supposed to be (https://github.com/WebAssembly/exception-handling/pull/175#discussion_r670152661).

My understanding was it should be valid and rethrow the exception out of the function, as discussed in #146 (see example there, especially for `delegate 3`) and as covered in the core tests [in this test case](https://github.com/WebAssembly/exception-handling/blob/master/test/core/try_delegate.wast#L48) which I'll just inline below:

```wasm
  (func (export "delegate-to-caller")
    (try (do (try (do (throw $e0)) (delegate 1))) (catch_all))
  )
```

There's an inconsistency though with the overview in https://github.com/WebAssembly/exception-handling/blob/master/proposals/exception-handling/Exceptions.md as it says:

> When the specified label within a delegate instruction does not correspond to a try instruction, it is a validation failure.

Assuming we agree on the behavior, we could change the overview to say "does not correspond to a try instruction or function body" (and an example could be helpful) and possibly adjust some other prose.

### PR DESCRIPTION
**Pull request by aheejin**
This adds the changes decided in #176 and adds a modified version from
https://github.com/WebAssembly/exception-handling/issues/146#issue-794482909
for clarification. (The example is not the same by the way; the `catch`
from the outermost `try` has been moved)

Closes #176.

### CODE REVIEW

File: proposals/exception-handling/Exceptions.md
Comment (rossberg) [inline proposals/exception-handling/Exceptions.md]: Nit: Delegate is no longer targeting catch after this change, it simply targets a block. I would perhaps rephrase this accordingly and say that all properties of Wasm's structured control flow labels apply, i.e., you can only target labels of outer blocks.

File: proposals/exception-handling/Exceptions.md
Comment (rossberg) [inline proposals/exception-handling/Exceptions.md]:
```suggestion
If `delegate` targets the implicit function body block, then in effect it delegates the
```

**Review by rossberg — APPROVED** [review]
Thanks, LGTM, merely some nits.

File: proposals/exception-handling/Exceptions.md
Comment (thibaudmichaud) [inline proposals/exception-handling/Exceptions.md]: I haven't seen this case mentioned explicitly yet, I think it fits here.
```suggestion
propagate to the outer scope and will be caught by the next matching
try-catches, or rethrown to the caller if there is no outer try block. In the examples, catches are annotated with `($label_name)` to
```

File: proposals/exception-handling/Exceptions.md
Comment (aheejin) [reply proposals/exception-handling/Exceptions.md]: True, but the example below is, the `delegate` targets `$l1`, which is outside of the instruction because the label is attached to the `try`, but because it is below `catch` so it ends up not targeting `$l1` but rather `$l0`, and this was what I tried to explain. Do you think simply describing that `delegate` target labels of outer blocks can explain this case?

File: proposals/exception-handling/Exceptions.md
Comment (takikawa) [inline proposals/exception-handling/Exceptions.md]: This is a minor nit, but I liked this better with the comment `;;` (for consistency with the rest of the doc).

File: proposals/exception-handling/Exceptions.md
Comment (takikawa) [inline proposals/exception-handling/Exceptions.md]: Should this be "greater than or equal to"? (or "depth of the outermost block" without +1?)

**Review by takikawa — APPROVED** [review]

File: proposals/exception-handling/Exceptions.md
Comment (aheejin) [reply proposals/exception-handling/Exceptions.md]: Done

File: proposals/exception-handling/Exceptions.md
Comment (aheejin) [reply proposals/exception-handling/Exceptions.md]: Hmm, yeah, "the depth of the outermost block" is confusing. I tried to rephrase it to
> If the immediate is greater than the number of block nesting including the implicit function-level block, it is a validation failure.

Do you think it's better?

Comment (takikawa) [reply proposals/exception-handling/Exceptions.md]: I think there's still some potential for confusion if someone interprets "number of block nesting" as counting from 1 and not 0. (In the example right above here you might count nestings of `try` and function body as 2, but immediates >=2 are not allowed here) Perhaps it could explicitly say it's counting from zero?

Comment (aheejin) [reply proposals/exception-handling/Exceptions.md]: Sorry, yeah, it should change to "greater than equal to". Intuitively the level of nesting is a number > 0 when there's any, and that's what I meant too. I changed it to
> If the immediate is greater than or equal to the number of block nesting including the implicit function-level block, it is a validation failure. In this example, any number equal to or greater than 2 is now allowed.

Comment (aheejin) [reply proposals/exception-handling/Exceptions.md]: Oh... now allowed -> not allowed...

File: proposals/exception-handling/Exceptions.md
Comment (aheejin) [reply proposals/exception-handling/Exceptions.md]: @rossberg Do you still think this needs fixing?

Comment (rossberg) [reply proposals/exception-handling/Exceptions.md]: I would say s.th like this:

"""
Like branches, `delegate` can only target outer blocks, and effectively rethrows the exception in that block. Consequently, delegating to a specific `catch` or `catch_all` handler requires targeting the respective label from within the associated `try` block. Delegating to a label from within a `catch` block does delegate the exception to the next enclosing handler -- analogous to performing a `throw` within a `catch` block, that handler is no longer active at that point.
"""

Comment (aheejin) [reply proposals/exception-handling/Exceptions.md]: Done

**Comment by aheejin** [comment]
@takikawa I'm not very familiar with what the status of newly added CI is, but is it expected to fail? This PR doesn't modify anything with the interpreter or the spec tests though.

**Comment by takikawa** [comment]
@aheejin Yes, it should be ok that these two things (ref interpreter & core spec) are failing, and it should be ok to just merge. Sorry for the confusion.

The issue with the interpreter CI is that the interpreter changes haven't been merged yet so the exception tests don't pass (BTW, did you or anyone else want to look at that PR more or should I merge it as it has one approval already?), but it should be ok after it's merged.

The core spec fails most likely because it's out of sync with the upstream spec. This should be fixed once it's rebased.

**Pull request merged**
